### PR TITLE
feat: add a simple cluster upgrade framework

### PR DESCRIPTION
**Pull request by csuzhangxc**
### What problem does this PR solve? 

part of #774.

### What is changed and how it works?

just add a simple cluster upgrade framework

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

 - Need to update the documentation

### CODE REVIEW

File: pkg/upgrade/upgrade.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+	// 1. get previous version from etcd.
+	preVer, _, err := GetVersion(cli)
+	if err != nil {
+		return err
+	}
+
+	// 2. check if any previous version exists.
+	if preVer.NotSet() {
+		// no initialization operations exist for Ver1 now,
+		// and any operations (may includes `upgrades`) if needed later.
```
Comment (lance6716) [inline pkg/upgrade/upgrade.go:43]:
```suggestion
		// add any operations (may includes `upgrades`) if needed later.
```

typo?

File: pkg/upgrade/upgrade.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+		// previous == current version, no need to upgrade.
+		return nil
+	} else if cmp > 0 {
+		// previous >= current version, this often means a older version of DM-master become the leader after started,
+		// do nothing for this now.
+		return nil
+	}
+
+	// 4. do upgrade operations.
+	for _, upgrade := range upgrades {
```
Comment (lance6716) [inline pkg/upgrade/upgrade.go:60]: if every update function has no side effects, it's OK not skipping out-of-time updates (like not skipping updateVer1 when 2.0 -> 3.0).

Comment (csuzhangxc) [reply pkg/upgrade/upgrade.go:60]: we may achieve _every update function has no side effects_, seems TiDB did this until now.

File: pkg/upgrade/version_test.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	c.Assert(ve.NotSet(), IsTrue)
+	c.Assert(CurrentVersion.NotSet(), IsFalse)
+
+	v1 := NewVersion(1, "v2.0.0")
+	v2 := NewVersion(2, "v2.1.0")
+	c.Assert(v1.Compare(v2), Equals, -1)
+	c.Assert(v2.Compare(v1), Equals, 1)
+
+	v11 := v1
+	c.Assert(v11.Compare(v1), Equals, 0)
```
Comment (lance6716) [inline pkg/upgrade/version_test.go:43]: how about testing "we only compare `InternalNo` now."

Comment (csuzhangxc) [reply pkg/upgrade/version_test.go:43]: tested in https://github.com/pingcap/dm/pull/839/commits/862ad50bbf329a7b3ce8b6ed056f5232d702b222.

**Comment by lance6716** [comment]
LGTM

**Pull request merged**